modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,691 @@
+        for (const auto& nh : nodes) {
+            if (m_gm.metadata(nh).get<NodeType>().t == NodeType::OP) {
+                garg_size += m_gm.metadata(nh).get<Op>().args.size();
+            }
+        }
+        m_oak_wrapped_args.reserve(garg_size);
+
+        size_t curr_arg_idx = 0;
+
+        // Create OAK node for each node in this backend
+        for (const auto& nh : nodes) {
+            if (m_gm.metadata(nh).get<NodeType>().t == NodeType::OP) {
+                const auto& op = m_gm.metadata(nh).get<Op>();
+                const auto &u = ConstOAKGraph(m_g).metadata(nh).get<OAKComponent>();
+                // pass kernel input args and compile args to prepare OAK node and
+                // store it to link later
+                m_oak_nodes[nh] = {};
+                m_oak_nodes[nh].inputs.reserve(op.args.size());
+                m_oak_nodes[nh].outputs.reserve(op.outs.size());
+
+                std::vector<std::reference_wrapper<cv::GArg>> in_ctx_args;
+                in_ctx_args.reserve(op.args.size());
+                for (auto &op_arg : op.args) in_ctx_args.push_back(packArg(op_arg,
+                                                                           m_oak_nodes[nh].inputs,
+                                                                           m_oak_wrapped_args));
+                for (auto &&op_out : op.outs) outArg(op_out, m_oak_nodes[nh].outputs);
+                GAPI_Assert(!m_oak_nodes[nh].inputs.empty());
+                GAPI_Assert(!m_oak_nodes[nh].outputs.empty());
+
+                GOAKContext ctx(m_pipeline, m_camera_wh, in_ctx_args, m_oak_nodes[nh].outputs);
+                m_oak_nodes[nh].node = u.k.m_f(ctx, m_in_queues);
+                GAPI_Assert(m_oak_nodes[nh].node != nullptr);
+
+                // Restore wrapped input references
+                // FIXME: where is a problem somewhere - should work w/o this hack
+                for (size_t i = 0; i < m_oak_nodes[nh].inputs.size(); ++i) {
+                    m_oak_nodes[nh].inputs[i] = m_oak_wrapped_args.at(curr_arg_idx + i).get<ExtractTypeHelper::InputPtr>();
+                }
+                curr_arg_idx += op.args.size();
+            }
+        }
+
+        // Check that all inputs and outputs are properly filled after constructing kernels
+        // to then link it together
+        for (const auto& node : m_oak_nodes) {
+            auto el = node.second;
+            if (std::any_of(el.inputs.cbegin(), el.inputs.cend(),
+                            [](ExtractTypeHelper::InputPtr ptr) {
+                    return ptr == nullptr;
+                })) {
+                GAPI_Assert(false && "DAI input are not set");
+            }
+            if (std::any_of(el.outputs.cbegin(), el.outputs.cend(),
+                            [](ExtractTypeHelper::OutputPtr ptr) {
+                    return ptr == nullptr;
+                })) {
+                GAPI_Assert(false && "DAI outputs are not set");
+            }
```

Comment:
You can make a lambda and reuse it for both input and output vectors

Reply:
Inputs and outputs have different types inside